### PR TITLE
[ENH] Offline batching

### PR DESCRIPTION
**Pull request by paxcema**
## Changelog
1. `EncodedDs.__getitem__` now retrieves rows from an "offline" cache that is populated at the end of the initialization method, instead of just-in-time. This enables encoders (which are assumed to be optimized, mostly there but some exceptions we're working on at the moment) to do batch encoding at a much faster rate.
2. Optimized some encoders with vectorized operations (see point 1): `NumericalEncoder`, `TsNumericEncoder`, `TsArrayNumericEncoder`
3. Reactivated Neural mixer's early stopping with a patience-based mechanism, and a learning rate search procedure with a standard range of options.
4. Deactivated RandomForest mixer's hyperparam optimization by default due to small loss in accuracy traded by a large speedup in training runtime.

## Effects
Lightwood becomes an order of magnitude faster across all ~50 benchmarked datasets (average speedup is ~7.5x, median is ~16.5x)

### CODE REVIEW

File: lightwood/encoder/numeric/numeric.py [hunk 1]
```diff
@@ -112,40 +95,32 @@ def decode(self, encoded_values: Union[List[Number], torch.Tensor], decode_log:
...
+                    np.isnan(vector[2]) or vector[2] == float('inf'):
 
-                real_value = vector[3] * self._abs_mean
+                log.error(f'Got weird target value to decode: {vector}')
+                real_value = pow(10, 63)
+
+            elif decode_log:
+                sign = -1 if vector[0] < 0.5 else 1
+                try:
+                    real_value = math.exp(vector[1]) * sign
```
Comment (pafluxa) [inline lightwood/encoder/numeric/numeric.py:114]:
```
real_value = numpy.exp(vector[:, 1]) * sign
overflowMask = vector[:, 1] >= 63
real_value[overflowMask] = 10**63 # pow() is way more expensive than **
```

File: lightwood/encoder/numeric/numeric.py [hunk 2]
```diff
@@ -112,40 +95,32 @@ def decode(self, encoded_values: Union[List[Number], torch.Tensor], decode_log:
...
+                except OverflowError:
+                    real_value = pow(10, 63) * sign
+            else:
```
Comment (pafluxa) [inline lightwood/encoder/numeric/numeric.py:117]:
```
# outside the for loop
real_value = vector[:, 2] * self._abs_mean
```

File: lightwood/encoder/numeric/numeric.py [hunk 3]
```diff
@@ -112,40 +95,32 @@ def decode(self, encoded_values: Union[List[Number], torch.Tensor], decode_log:
...
-                if self._type == 'int':
-                    real_value = round(real_value)
+            if self.positive_domain:
+                real_value = abs(real_value)
 
-            if isinstance(real_value, torch.Tensor):
-                real_value = real_value.item()
+            # if isinstance(real_value, torch.Tensor):
+            #     real_value = real_value.item()
             ret.append(real_value)
```
Comment (pafluxa) [inline lightwood/encoder/numeric/numeric.py:125]:
```
validMask = ˜numpy.logical_and(overflowMask, maskNone)
ret[validMask] = real_value[validMask]
```

File: lightwood/mixer/neural.py [hunk 4]
```diff
@@ -106,32 +112,36 @@ def _select_criterion(self) -> torch.nn.Module:
...
         stop = False
-        batches = 0
-        for epoch in range(1, 101):
-            if stop:
-                break
 
-            for i, (X, Y) in enumerate(dl):
-                if stop:
-                    break
+        _, test_batch = next(enumerate(dl))
```
Comment (pafluxa) [inline lightwood/mixer/neural.py:131]: You might want to replace `enumerate()` by `iter()`

**Review by pafluxa — COMMENTED** [review]
Review completed. Please see comments for details on suggestions.

I would think that the only critical modifications should be carried out on `encoded_ds.py`, as the current implementation of the caching mechanism will produce unwanted results on the `__getitem__()` method. Other than that, the modifications look OK. 

Nicely done!!

File: lightwood/data/encoded_ds.py [hunk 5]
```diff
@@ -56,21 +59,28 @@ def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor]:
             if self.cache[idx] is not None:
                 return self.cache[idx]
 
+        X, Y = self._encode_idxs(idx)
+
+        if self.cache_encoded:
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/data/encoded_ds.py:64]: This was a total oversight on my end :grin: I somehow ate the second return while reformatting a few things, it seems. Will do!

File: lightwood/data/encoded_ds.py [hunk 6]
```diff
@@ -33,6 +33,9 @@ def __init__(self, encoders: List[BaseEncoder], data_frame: pd.DataFrame, target
                                            self.input_length + self.encoders[col].output_size)
                 self.input_length += self.encoders[col].output_size
 
+        # if cache enabled, we immediately build it
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/data/encoded_ds.py:36]: Absolutely. The flag actually exists (`self.cache_encoded`, and I do think it is best to rename it to `use_cache` or something similar). I am inclined to keep using the cache by default, though (it is the currently active method).

File: lightwood/encoder/numeric/numeric.py [hunk 7]
```diff
@@ -43,55 +45,36 @@ def prepare(self, priming_data: Iterable):
...
-                    log.error(f'Can\'t encode input value: {real}, exception: {e}')
-
-            ret.append(vector)
-
+        # todo: wrap with try/except to cover non-real edge cases
+        if not self.positive_domain:
+            sign = np.vectorize(lambda x: 0 if x < 0 else 1)(data)
+        else:
+            sign = np.zeros(len(data))
+        log_value = np.vectorize(lambda x: math.log(abs(x)) if abs(x) > 0 else -20)(data)
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/encoder/numeric/numeric.py:64]: I like the idea of exploring ported encoders (or part of them, anyway) to e.g. polars, cudf, or as C extensions as you said. Let's do it after we do a first initial optimization pass :muscle:

File: lightwood/encoder/numeric/numeric.py [hunk 8]
```diff
@@ -43,55 +45,36 @@ def prepare(self, priming_data: Iterable):
...
+            components = [sign, log_value, exp]
+        else:
+            # todo: if can't encode return 0s and log.error(f'Can\'t encode input value: {real}, exception: {e}')
+            nones = np.vectorize(lambda x: 1 if is_none(x) else 0)(data)
+            components = [sign, log_value, exp, nones]
+
+        ret = torch.Tensor(np.array(components)).T
         return torch.Tensor(ret)
 
     def decode(self, encoded_values: Union[List[Number], torch.Tensor], decode_log: bool = None) -> list:
```
[the remark this answers is not in the document]
Comment (paxcema) [reply lightwood/encoder/numeric/numeric.py:80]: This works perfectly! I've opted for `full_like(fill_value=np.finfo(np.float64).max` assuming it's a _tad_ faster, probably insignificant though.
